mls_validation_service/src/handlers.rs

```diff
@@ -0,0 +1,256 @@
+use openmls_rust_crypto::OpenMlsRustCrypto;
+use openmls_traits::OpenMlsProvider;
+use tonic::{Request, Response, Status};
+
+use openmls::{
+    prelude::{KeyPackageIn, MlsMessageIn, ProtocolMessage, TlsDeserializeTrait},
+    versions::ProtocolVersion,
+};
+use xmtp_proto::xmtp::mls_validation::v1::{
+    validate_group_messages_response::ValidationResponse as ValidateGroupMessageValidationResponse,
+    validate_key_packages_response::ValidationResponse as ValidateKeyPackageValidationResponse,
+    validation_api_server::ValidationApi, ValidateGroupMessagesRequest,
+    ValidateGroupMessagesResponse, ValidateKeyPackagesRequest, ValidateKeyPackagesResponse,
+};
+
+use crate::validation_helpers::{hex_encode, identity_to_wallet_address};
+
+#[derive(Debug, Default)]
+pub struct ValidationService {}
+
+#[tonic::async_trait]
+impl ValidationApi for ValidationService {
+    async fn validate_key_packages(
+        &self,
+        request: Request<ValidateKeyPackagesRequest>,
+    ) -> Result<Response<ValidateKeyPackagesResponse>, Status> {
+        let out: Vec<ValidateKeyPackageValidationResponse> = request
+            .into_inner()
+            .key_packages
+            .into_iter()
+            .map(
+                |kp| match validate_key_package(kp.key_package_bytes_tls_serialized) {
+                    Ok(res) => ValidateKeyPackageValidationResponse {
+                        installation_id: res.installation_id,
+                        wallet_address: res.wallet_address,
+                        error_message: "".to_string(),
+                        is_ok: true,
+                    },
+                    Err(e) => ValidateKeyPackageValidationResponse {
+                        is_ok: false,
+                        error_message: e,
+                        installation_id: "".to_string(),
+                        wallet_address: "".to_string(),
+                    },
+                },
+            )
+            .collect();
+
+        Ok(Response::new(ValidateKeyPackagesResponse {
+            responses: out,
+        }))
+    }
+
+    async fn validate_group_messages(
+        &self,
+        request: Request<ValidateGroupMessagesRequest>,
+    ) -> Result<Response<ValidateGroupMessagesResponse>, Status> {
+        let out: Vec<ValidateGroupMessageValidationResponse> = request
+            .into_inner()
+            .group_messages
+            .into_iter()
+            .map(|message| {
+                match validate_group_message(message.group_message_bytes_tls_serialized) {
+                    Ok(res) => ValidateGroupMessageValidationResponse {
+                        group_id: res.group_id,
+                        epoch: res.epoch,
```

Comment:
This would be used for proposals/commits too right? If so, is this the epoch going forward, or the epoch the message was expected to be sent on - don't we need both?

And is the golang server going to validate that the current epoch matches the expected epoch outside of this one?

Reply:
This will get used for any group message (application, proposal, commit).

We don't do any validation on the epoch. I probably should just remove the field from the response. It predates our convo with @Bren2010 around relying on total ordering rather than epoch validation.